### PR TITLE
Refactor Folder Domain

### PR DESCRIPTION
**Pull request by Hinton**
## Type of change

- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective

Suggestion on how we can refactor our domain models to avoid weird chains (`Response -> Data -> Domain -> View`). This puts the `Domain` in the center with everything converting to/from it.

The new data model is represented by the following objects:

* Domain: Our domain model
* Data: Our long term local storage model
* Decrypted: Decrypted Domain
* Request/Response: Servers representation

Something I haven't gotten around to investigating is how we can actually store `Domain` in memory instead of converting back/from Data all the time.

I also deprecated `EncString.decrypt` and replaced it with `EncString.decryptWithCryptoService` which uses a local crypto service instead of the global container.

## Before you submit

- [ ] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

**Comment by Hinton** [comment]
I'm still a bit unhappy about `FolderView.fromFolder`, but I prefer that over having `Domain` convert itself to `View` since that would make a circular dependency, `view -> domain -> view`.

File: common/src/abstractions/folder.service.ts [hunk 1]
```diff
@@ -1,21 +1,19 @@
...
   clearCache: (userId?: string) => Promise<void>;
-  encrypt: (model: FolderView, key?: SymmetricCryptoKey) => Promise<Folder>;
   get: (id: string) => Promise<Folder>;
   getAll: () => Promise<Folder[]>;
   getAllDecrypted: () => Promise<FolderView[]>;
   getAllNested: () => Promise<TreeNode<FolderView>[]>;
   getNested: (id: string) => Promise<TreeNode<FolderView>>;
   saveWithServer: (folder: Folder) => Promise<any>;
-  upsert: (folder: FolderData | FolderData[]) => Promise<any>;
+  upsert: (folder: FolderData) => Promise<any>;
```
Comment (MGibson1) [inline common/src/abstractions/folder.service.ts:14]: Are you planning on an `upsertMany` if a bulk operation is supported?

File: common/src/models/domain/encString.ts [hunk 2]
```diff
@@ -96,11 +96,10 @@ export class EncString {
...
   }
 
+  /**
+   * @deprecated As of 2022-04-20, use {@link decryptWithCryptoService} instead
+   */
   async decrypt(orgId: string, key: SymmetricCryptoKey = null): Promise<string> {
-    if (this.decryptedValue != null) {
-      return this.decryptedValue;
-    }
-
```
Comment (MGibson1) [inline common/src/models/domain/encString.ts:103]: I don't think it's a good idea to remove this guard.

Are you thinking that the object hierarchy makes multiple decrypts less likely?

File: common/src/models/domain/folder.ts [hunk 3]
```diff
@@ -1,40 +1,7 @@
-import { FolderData } from "../data/folderData";
-import { FolderView } from "../view/folderView";
-
-import Domain from "./domainBase";
 import { EncString } from "./encString";
 
-export class Folder extends Domain {
+export interface Folder {
```
Comment (MGibson1) [inline common/src/models/domain/folder.ts:3]: 🥳

File: common/src/models/export/folderExport.ts [hunk 4]
```diff
@@ -14,7 +14,7 @@ export class FolderExport {
     return view;
   }
 
-  static toDomain(req: FolderExport, domain = new FolderDomain()) {
+  static toDomain(req: FolderExport, domain: FolderDomain = {} as FolderDomain) {
```
Comment (MGibson1) [inline common/src/models/export/folderExport.ts:17]: Is this due to the interface?

File: common/src/services/folder.service.ts [hunk 5]
```diff
@@ -30,33 +30,25 @@ export class FolderService implements FolderServiceAbstraction {
     await this.stateService.setDecryptedFolders(null, { userId: userId });
   }
 
-  async encrypt(model: FolderView, key?: SymmetricCryptoKey): Promise<Folder> {
-    const folder = new Folder();
-    folder.id = model.id;
-    folder.name = await this.cryptoService.encrypt(model.name, key);
-    return folder;
+  async decrypt(model: Folder, key?: SymmetricCryptoKey): Promise<FolderView> {
```
Comment (MGibson1) [inline common/src/services/folder.service.ts:33]: I don't think this is used externally

**Review by MGibson1 — COMMENTED** [review]
I have a few questions, but the primary concern is I don't yet understand what this will be cleaning up.

I **do** want a better definition of all of the models we have and a definitive hierarchy sounds great, but it didn't result in the code cleanup I expected.

File: common/src/models/domain/encString.ts (hunk 2 above)
Comment (Hinton) [reply common/src/models/domain/encString.ts:103]: It's called in `decryptWithCryptoService`. So the guard is still there just moved below the container service initialization.

File: common/src/models/export/folderExport.ts (hunk 4 above)
Comment (Hinton) [reply common/src/models/export/folderExport.ts:17]: Yes

File: common/src/services/folder.service.ts (hunk 5 above)
Comment (Hinton) [reply common/src/services/folder.service.ts:33]: I meant to remove that 😢

File: common/src/abstractions/folder.service.ts (hunk 1 above)
Comment (Hinton) [reply common/src/abstractions/folder.service.ts:14]: Yes, I think that's a cleaner pattern than allowing two different data models.

`upsert(folder) => upsertMany([folder])`

**Pull request closed without merging**